# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/AnimatedPropertyID.h
# Commit: 63603267402e
# Full Hash: 63603267402ee93d6551ef60c35191d99aaee06b
# Author: Zach Hoffman <zach@zrhoffman.net>
# Date: 2023-12-18 17:17:57
# Regressor Bug: 1846516
# File Overlap Count: 3
# Description:
#   Bug 1846516 - [css-properties-values-api] Introduce AnimatedPropertyID/AnimatedPropertIDSet. r=firefox-animation-reviewers,firefox-style-system-reviewers,layout-reviewers,emilio,hiro
#   
#   This will make possible to animate custom properties on Gecko side. For now, the
#   animation code keeps only dealing with nsCSSPropertyID, so behavior is unchanged.
#   
# ==============================================================================

diff -r 753bc39f59ff -r 63603267402e layout/style/AnimatedPropertyID.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/layout/style/AnimatedPropertyID.h	Mon Dec 18 09:24:08 2023 +0000
@@ -0,0 +1,63 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_AnimatedPropertyID_h
+#define mozilla_AnimatedPropertyID_h
+
+#include "nsCSSPropertyID.h"
+#include "nsCSSProps.h"
+#include "nsString.h"
+#include "mozilla/HashFunctions.h"
+#include "mozilla/ServoBindings.h"
+
+namespace mozilla {
+
+struct AnimatedPropertyID {
+  explicit AnimatedPropertyID(nsCSSPropertyID aProperty) : mID(aProperty) {
+    MOZ_ASSERT(aProperty != eCSSPropertyExtra_variable,
+               "Cannot create an AnimatedPropertyID from only a "
+               "eCSSPropertyExtra_variable.");
+  }
+
+  explicit AnimatedPropertyID(RefPtr<nsAtom> aCustomName)
+      : mID(eCSSPropertyExtra_variable), mCustomName(std::move(aCustomName)) {}
+
+  nsCSSPropertyID mID = eCSSProperty_UNKNOWN;
+  RefPtr<nsAtom> mCustomName;
+
+  bool IsCustom() const { return mID == eCSSPropertyExtra_variable; }
+  bool operator==(const AnimatedPropertyID& aOther) const {
+    return mID == aOther.mID &&
+           (!IsCustom() || mCustomName == aOther.mCustomName);
+  }
+  bool operator!=(const AnimatedPropertyID& aOther) const {
+    return !(*this == aOther);
+  }
+  bool IsValid() const {
+    if (mID == eCSSProperty_UNKNOWN) {
+      return false;
+    }
+    return IsCustom() == !!mCustomName;
+  }
+
+  void ToString(nsAString& aString) const {
+    if (IsCustom()) {
+      MOZ_ASSERT(mCustomName, "Custom property should have a name");
+      mCustomName->ToString(aString);
+    } else {
+      aString.Assign(NS_ConvertUTF8toUTF16(nsCSSProps::GetStringValue(mID)));
+    }
+  }
+
+  HashNumber Hash() const {
+    HashNumber hash = mCustomName ? mCustomName->hash() : 0;
+    return AddToHash(hash, mID);
+  }
+};
+
+}  // namespace mozilla
+
+#endif  // mozilla_AnimatedPropertyID_h